File: trunk/java/USB2XXX/src/USB2XXX_LIN_Test.java
import com.usbxyz.USB_Device;
import com.usbxyz.USB2LIN;

public class USB2XXX_LIN_Test {
	  /** 
     * Launch the application. 
     */  
    public static void main(String[] args) {   
        int ret;
        int DevHandle = 0;
        int[] DevHandleArry = new int[20];
        boolean state;
        int LINMasterIndex = 0;
        int LINSlaveIndex = 0;
        byte[] WriteDataBuffer = new byte[1024];
        byte[] ReadDataBuffer = new byte[1024];
        //扫描设备
        ret = USB_Device.INSTANCE.USB_ScanDevice(DevHandleArry);
        if(ret > 0){
        	System.out.println("Device Num = "+ret);
        	DevHandle = DevHandleArry[0];
        }else{
        	System.out.println("No device");
        	return;
        }
        //打开设备
        state = USB_Device.INSTANCE.USB_OpenDevice(DevHandle);
        if(!state){
        	System.out.println("open device error");
        	return;
        }
        //获取设备信息
        USB_Device.DEVICE_INFO DevInfo = new USB_Device.DEVICE_INFO();
        byte[] funcStr = new byte[128];
        state = USB_Device.INSTANCE.USB_GetDeviceInfo(DevHandle,DevInfo,funcStr);
        if(!state){
        	System.out.println("get device infomation error");
        	return;
        }else{
            try {
            	System.out.println("Firmware Info:");
            	System.out.println("--Name:" + new String(DevInfo.FirmwareName, "UTF-8"));
            	System.out.println("--Build Date:" + new String(DevInfo.BuildDate, "UTF-8"));
            	System.out.println(String.format("--Firmware Version:v%d.%d.%d", (DevInfo.FirmwareVersion >> 24) & 0xFF, (DevInfo.FirmwareVersion >> 16) & 0xFF, DevInfo.FirmwareVersion & 0xFFFF));
            	System.out.println(String.format("--Hardware Version:v%d.%d.%d", (DevInfo.HardwareVersion >> 24) & 0xFF, (DevInfo.HardwareVersion >> 16) & 0xFF, DevInfo.HardwareVersion & 0xFFFF));
            	System.out.println("--Functions:" + new String(funcStr, "UTF-8"));
            } catch (Exception ep) {
                ep.printStackTrace();
            }
        }
        //初始化配置LIN为主机模式
        USB2LIN.LIN_CONFIG LINConfig = new USB2LIN.IIC_CONFIG();
        LINConfig.BaudRate = 19200;
        LINConfig.BreakBits = USB2LIN.LIN_BREAK_BITS_11;
        LINConfig.CheckMode = USB2LIN.LIN_CHECK_MODE_EXT;
        LINConfig.MasterMode = USB2LIN.LIN_MASTER;
        ret = LIN_Init(DevHandle,LINMasterIndex,LINConfig);
        if(ret != LIN_SUCCESS){
            System.out.println("Config LIN failed!");
            return ;
        }else{
            System.out.println("Config LIN Success!");
        }
        //主机写数据
        for(int ID=0;ID<10;ID++){
            USB2LIN.LIN_MSG msg = new USB2LIN.LIN_MSG();
            for(int i=0;i<8;i++){
                msg.Data[i] = ID+i;
            }
            msg.DataLen = 8;
            msg.ID = ID;
            ret = LIN_Write(DevHandle,LINMasterIndex,msg,1);
            if(ret != LIN_SUCCESS){
                System.out.println("LIN write data failed!");
                return;
            }else{
                System.out.println("LIN write data success!");
            }
        }
        //主机读数据
        for(int ID=0;ID<10;ID++){
            USB2LIN.LIN_MSG msg = new USB2LIN.LIN_MSG();
            msg.ID = ID;
            ret = LIN_Read(DevHandle,LINMasterIndex,msg,1);
            if(ret < LIN_SUCCESS){
                System.out.println("LIN read data failed!");
            }else{
                System.out.println("Master LIN Read: ID = 0x%02X Data = ",ID);
                for(int i=0;i<msg.DataLen;i++){
                    System.out.print("0x%02X ",msg.Data[i]);
                }
                System.out.print("\n");
            }
        }
        //初始化配置LIN为从机模式
        USB2LIN.LIN_CONFIG LINConfig = new USB2LIN.IIC_CONFIG();
        LINConfig.BaudRate = 19200;
        LINConfig.BreakBits = USB2LIN.LIN_BREAK_BITS_11;
        LINConfig.CheckMode = USB2LIN.LIN_CHECK_MODE_EXT;
        LINConfig.MasterMode = USB2LIN.LIN_SLAVE;
        ret = LIN_Init(DevHandle,LINSlaveIndex,LINConfig);
        if(ret != LIN_SUCCESS){
            System.out.println("Config LIN failed!");
            return ;
        }else{
            System.out.println("Config LIN Success!");
        }
        //设置从机模式下所有ID都为从接收数据模式，这样就可以获取到主机发送过来的所有数据
        USB2LIN.LIN_MSG[] LINSlaveData = new USB2LIN.LIN_MSG().toArray(64);
        for(int i=0;i<64;i++){
            LINSlaveData[i].DataLen = 9;//最大8Byte数据+1Byte和校验
            LINSlaveData[i].ID = i;		//ID值
        }
        ret = LIN_SlaveSetIDMode(DevHandle,LINSlaveIndex,USB2LIN.LIN_SLAVE_READ,LINSlaveData,64);
        if(ret != LIN_SUCCESS){
            System.out.println("Set LIN operation mode failed!");
            return;
        }else{
            System.out.println("Set LIN operation mode success!");
        }
        //主机发送数据之后，可以读取从机接收到的数据
        USB2LIN.LIN_MSG[] LINSlaveDataBuffer = new USB2LIN.LIN_MSG[1024];//为了防止缓冲区溢出，可以将接收数据缓冲区设置大一点
        ret = LIN_SlaveGetData(DevHandle,LINSlaveIndex,LINSlaveDataBuffer);
        if(ret < LIN_SUCCESS){
            System.out.println("LIN slave read data error!");
            return;
        }else if(ret == 0){
            System.out.println("LIN slave read no data!");
        }else{
            System.out.println("LIN slave read data:");
            for(int i=0;i<ret;i++){
                System.out.print("Data[%d]: ",i);
                System.out.print("ID = 0x%02X ",LINSlaveDataBuffer[i].ID);
                System.out.print("Data = ");
                for(int j=0;j<LINSlaveDataBuffer[i].DataLen;j++){
                    System.out.print("0x%02X ",LINSlaveDataBuffer[i].Data[j]);
                }
                System.out.print("\n");
            }
        }
        System.out.println("");
        System.out.println("USB2LIN test success!");
        //关闭设备
        USB_Device.INSTANCE.USB_CloseDevice(DevHandle);
    }  
}

File: trunk/vb.net/USB2LIN/USB2XXX_LIN_Test/USB2XXX_LIN_Test/USB2XXX_LIN_Test/USB2XXX_LIN_Test.vb

#Const LIN_MASTER = True

Module USB2XXX_LIN_Test

    Sub Main()
        Dim ret As Int32
        Dim State As Boolean
        Dim DeviceHandle(20) As UInt32
        Dim LINIndex As Byte
        LINIndex = 0
        ret = USB_ScanDevice(DeviceHandle)
        If ret <= 0 Then
            Console.WriteLine("No device connected")
        End If

        State = USB_OpenDevice(DeviceHandle(0))
        If State Then
            Console.WriteLine("Open device success!")
        Else
            Console.WriteLine("Open device failed!")
            Return
        End If
        '初始化配置LIN
        Dim LINConfig As usb2lin.LIN_CONFIG
        LINConfig.BaudRate = 19200
        LINConfig.BreakBits = usb2lin.LIN_BREAK_BITS_11
        LINConfig.CheckMode = usb2lin.LIN_CHECK_MODE_EXT
#If LIN_MASTER Then
        LINConfig.MasterMode = usb2lin.LIN_MASTER
#Else
        LINConfig.MasterMode = USB2LIN.LIN_SLAVE
#End If
        ret = usb2lin.LIN_Init(DeviceHandle(0), LINIndex, LINConfig)
        If ret <> usb2lin.LIN_SUCCESS Then
            Console.WriteLine("Config LIN failed!")
            Return
        Else
            Console.WriteLine("Config LIN success!")
        End If

#If LIN_MASTER Then
        '主机写数据
        Dim ID As Byte
        For ID = 0 To 9
            Dim msg(1) As usb2lin.LIN_MSG
            Dim i As Byte
            ReDim msg(0)
            ReDim msg(0).Data(8)
            For i = 0 To 7
                msg(0).Data(i) = ID + i
            Next
            msg(0).DataLen = 8
            msg(0).ID = ID
            ret = usb2lin.LIN_Write(DeviceHandle(0), LINIndex, msg, 1)
            If ret <> usb2lin.LIN_SUCCESS Then
                Console.WriteLine("LIN write data failed!")
                Return
            Else
                Console.WriteLine("LIN write data success!")
            End If
            '延时
            Threading.Thread.Sleep(10)
        Next ID
        '主机读数据
        For ID = 10 To 15
            Dim msg(1) As usb2lin.LIN_MSG
            ReDim msg(0)
            msg(0).ID = ID
            ReDim msg(0).Data(8)
            ret = usb2lin.LIN_Read(DeviceHandle(0), LINIndex, msg, 1)
            If ret < usb2lin.LIN_SUCCESS Then
                Console.WriteLine("LIN write data failed!")
                Return
            Else
                Console.Write("Master LIN Read: ID = 0x{0:X2} Data = ", msg(0).ID)
                If msg(0).DataLen > 0 Then
                    Dim i As Byte
                    For i = 0 To msg(0).DataLen - 1
                        Console.Write("0x{0:X2} ", msg(0).Data(i))
                    Next i
                End If
                Console.WriteLine("")
            End If
            '延时
            Threading.Thread.Sleep(10)
        Next ID
#Else
        '设置从机模式下所有ID都为从接收数据模式，这样就可以获取到主机发送过来的所有数据,初始化配置为从机后，默认所有ID都为接收数据模式，所以若是监听LIN总线数据，这个函数可以不用调用
        Dim LINSlaveData(64) As usb2lin.LIN_MSG
        Dim i As Int32
        For i = 0 To 63
            ReDim LINSlaveData(i)
            LINSlaveData(i).DataLen = 9 '最大8Byte数据+1Byte和校验
            LINSlaveData(i).ID = i      'ID值
            ReDim LINSlaveData(i).Data(8)
        Next i
        ret = usb2lin.LIN_SlaveSetIDMode(DeviceHandle(0), LINIndex, usb2lin.LIN_SLAVE_READ, LINSlaveData, 64)
        If ret <> usb2lin.LIN_SUCCESS Then
            Console.WriteLine("Set LIN operation mode failed!")
            Return
        Else
            Console.WriteLine("Set LIN operation mode success!")
        End If
        '从机接收数据,若要连续不断的监控LIN总线数据，可以循环调用LIN_SlaveGetData函数
        Dim LINSlaveDataBuffer(1024) As usb2lin.LIN_MSG '为了防止缓冲区溢出，可以将接收数据缓冲区设置大一点
        For i = 0 To LINSlaveDataBuffer.Length - 1
            ReDim LINSlaveDataBuffer(i)
            ReDim LINSlaveDataBuffer(i).Data(8)
        Next

        While True
            ret = usb2lin.LIN_SlaveGetData(DeviceHandle(0), LINIndex, LINSlaveDataBuffer)
            If ret < usb2lin.LIN_SUCCESS Then
                Console.WriteLine("LIN slave read data error!")
                Return
            ElseIf ret > 0 Then
                Console.WriteLine("LIN slave read data:")
                For i = 0 To ret - 1
                    Console.Write("Data[{0}]: ", i)
                    Console.Write("ID = 0x{0:X2} ", LINSlaveDataBuffer(i).ID)
                    Console.Write("Data = ")
                    Dim j As Int32
                    For j = 0 To LINSlaveDataBuffer(i).DataLen - 1
                        Console.Write("0x{0:X2} ", LINSlaveDataBuffer(i).Data(j))
                    Next j
                Next i
                Console.WriteLine("")
            Else
                'Console.WriteLine("LIN slave read no data!")
                '延时
                Threading.Thread.Sleep(10)
            End If
        End While
#End If

    End Sub

End Module
